get hardening policies for `replicasets`, `statefulsets` & `daemonsets`

Prateeknandle opened the pull request:
fixes #670 

changes:
- getting `replicasets`, `daemonsets` and `statefulsets` from k8s client and generating hardening policies respectively
- updating `clusterRole` for the same
- added test cases to test whether we get the hardening policy or not

Prateeknandle committed:
generating policy for replica-set & statefulset

Signed-off-by: Prateeknandle <prateeknandle@gmail.com>

vishnusomank commented:
LGTM! 

@Prateeknandle  Can we update the helm chart as well to match the `ClusterRole` ?

Prateeknandle commented:
> @Prateeknandle Can we update the helm chart as well to match the `ClusterRole` ?

updated

seswarrajan reviewed (approved):
LGTM

Prateeknandle commented:
can we merge the pr?

nyrahul commented:
> can we merge the pr?

The test is failing!

nyrahul commented:
Can we add a testcase to test this?

Prateeknandle commented:
> Can we add a testcase to test this?

tests for `recommend` are defined in `kubearmor-client`

> The test is failing!

it's because the tests are flaky

For testing whether changes are working or not we can see logs

vishnusomank reviewed (approved):
LGTM! 

Tested by deploying via deployment.yaml and helm. The code changes are able to detect and generate policies for Statefulset and Replicaset. 
Only issue is if a new replicaset or statefulset is deployed we'll have to wait for the cronjob to trigger (every 1 hour) to get hardening policies for those.

Prateeknandle commented:
With the new changes I've checked locally, tests are passing and discovery-engine is able to get the hardening policies.

the ginkgo test which are failing are the flaky ones and not related to this pr.

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,46 @@ func RecommendPolicyMain() {
 			}
 		}
 	}
+	for _, pol := range policy {
```
we don't need to loop through policy to append it. We can use something like
```
policies = append(policies, policy...)
```

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,46 @@ func RecommendPolicyMain() {
+			}
+		}
+	}
+	for _, pol := range policy {
```
same as the above comment

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,46 @@ func RecommendPolicyMain() {
+			}
+		}
+	}
+	for _, pol := range policy {
```
same as the above comment

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,46 @@ func RecommendPolicyMain() {
+		}
+	}
+	for _, pol := range policy {
+		policies = append(policies, pol)
```
We are not using this `policies` anywhere after it is returned right? Do we need this policies for any operation?

Prateeknandle replied:
We are using [GetHardenPolicy()](https://github.com/accuknox/discovery-engine/pull/681/files#diff-507e0011278270a66cc4717d5f666421567b9ca156b62c4c01670f52e6867a44R225) to get the policies for testing, RecommendPolicyMain() will directly not return it therefore using this func.

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,46 @@ func RecommendPolicyMain() {
+		return nil
 	}
 	systempolicy.UpdateSysPolicies(policies)
+	return policies
```
If we are returning the policies array from this function we can remove the `systempolicy.UpdateSysPolicies(policies)` from this function and use it in `RecommendPolicyMain()` so that it'll only get called once after we have all the policies for deployment, statefulset and replicasets.

Prateeknandle replied:
updated please check

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -135,6 +142,16 @@ func RecommendPolicyMain() {
 		log.Error().Msg(err.Error())
 		return
 	}
+	replicaSets, err := client.AppsV1().ReplicaSets("").List(context.Background(), metav1.ListOptions{})
```
@Prateeknandle When we get all the replicasets there is a possibility that the replicaset was created by a deployment. If that is the case then we will receive duplicate policies for the same. 
For eg., If we deploy wordpress-mysql application, we can see that it'll create replicasets from the deployment. 
```
k -n wordpress-mysql get deploy,rs                                                                                  (get_hardening_pol↓1|✚1)
NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/mysql       0/1     1            0           6d3h
deployment.apps/wordpress   1/1     1            1           6d3h

NAME                                   DESIRED   CURRENT   READY   AGE
replicaset.apps/mysql-76ddc6ddc4       1         1         0       6d3h
replicaset.apps/wordpress-787f45786f   1         1         1       6d3h
```
Naturally we wouldn't want policies to be generated for both (ie, duplicate) Can we handle this scenario as well?

Prateeknandle replied:
Handled in new commit

vishnusomank commented on tests/recommend/recommned_test.go:
```diff
@@ -0,0 +1,85 @@
+				if p.Metadata["namespace"] == "rs-demo" && p.Spec.Selector.MatchLabels["app"] == "nginx" {
+					flag = true
+				}
+				if flag == true {
```
we can use the break in the first if condition itself. 
```
if p.Metadata["namespace"] == "rs-demo" && p.Spec.Selector.MatchLabels["app"] == "nginx" {
    flag = true
    break
}
```

Prateeknandle replied:
Handled in new commit

vishnusomank commented on tests/recommend/recommned_test.go:
```diff
@@ -0,0 +1,81 @@
+
+	// create new namespace
+	_, err := util.Kubectl(fmt.Sprintf("delete namespace rs-demo"))
+	Expect(err).To(BeNil())
```
can we have both the sts and rs on the same namespace and make sure that they are deleted after the test is finished?
Currently the sts is not getting deleted after test finishes. @Prateeknandle

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,40 @@ func RecommendPolicyMain() {
+	for _, r := range replicaSets.Items {
+		for _, ns := range NsNotFilter {
+			if r.Namespace != ns && len(r.ObjectMeta.OwnerReferences) == 0 {
+				policy = generateHardenPolicy(r.Name, r.Namespace, r.Spec.Template.Labels)
```
Assume there are multiple namespaces with statefulset. When this for loop ends it will contain the harden `policy` only for the last namespace. All other namespaces hardening policies will be lost/overwritten. Am I missing something here?

Prateeknandle replied:
thanks for looking into it, updated the pr, please check.

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -135,6 +142,16 @@ func RecommendPolicyMain() {
 	}
+	replicaSets, err := client.AppsV1().ReplicaSets("").List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		log.Error().Msg(err.Error())
```
```suggestion
		log.Error().Msg("error getting replicasets err=" + err.Error())
```

Prateeknandle replied:
these are also addressed

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -135,6 +142,16 @@ func RecommendPolicyMain() {
+	}
+	statefulSets, err := client.AppsV1().StatefulSets("").List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		log.Error().Msg(err.Error())
```
```suggestion
		log.Error().Msg("error getting statefulsets err=" + err.Error())
```

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -143,22 +160,42 @@ func RecommendPolicyMain() {
+		for _, ns := range NsNotFilter {
 			if d.Namespace != ns {
-				generateHardenPolicy(d.Name, d.Namespace, d.Spec.Template.Labels)
+				policy = generateHardenPolicy(d.Name, d.Namespace, d.Spec.Template.Labels)
```
since we are not returning any error we can directly append the return value to `policies` like:
```
policies = append(policies, generateHardenPolicy(d.Name, d.Namespace, d.Spec.Template.Labels)...)
```

Prateeknandle replied:
updated

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -52,6 +52,11 @@ type LabelMap = map[string]string
+// NsNotFilter stores the namespace which needs to be filtered out
+var NsNotFilter = cfg.CurrentCfg.ConfigSysPolicy.NsNotFilter
+
+var policies []types.KnoxSystemPolicy
```
This is a very bad practice to keep the variables global. The functions using `policies` can no more be used in multi-threaded envs.

`policies` needs to be local vars.

Prateeknandle replied:
handled the bad func, PTAL

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -95,6 +101,7 @@ func StopRecommendWorker() {
 func StartRecommendCronJob() {
 	// init cron job
 	RecommendCronJob = cron.New()
+	NsNotFilter = cfg.CurrentCfg.ConfigSysPolicy.NsNotFilter
```
Why is this done multiple times? I see that NsNotFilter is already initialized at line 78.

vishnusomank commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -122,43 +129,37 @@ func StopRecommendCronJob() {
 // RecommendPolicyMain generates recommended policies from policy-template GH
 func RecommendPolicyMain() {
 
-	nsNotFilter := cfg.CurrentCfg.ConfigSysPolicy.NsNotFilter
```
can we have the `nsNotFilter` string created in this function as a local variable and pass it to `GetHardenPolicy()` rather than having it as a Global Variable

Prateeknandle replied:
added

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -180,3 +176,43 @@ func uniqueNsDeploy(deployName, deployNamespace string) *types.Deployment {
+	var policies []types.KnoxSystemPolicy
+	if !isLatest() {
+		if _, err := DownloadAndUnzipRelease(); err != nil {
+			log.Error().Msgf("Unable to download %v", err.Error())
```
Should this continue if the operation fails?

nyrahul replied:
It would be good to print the version returned by `DownloadAndUnzipRelease` so as to track.

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -180,3 +176,43 @@ func uniqueNsDeploy(deployName, deployNamespace string) *types.Deployment {
+
+	for _, r := range replicaSets.Items {
+		for _, ns := range nsNotFilter {
+			if r.Namespace != ns && len(r.ObjectMeta.OwnerReferences) == 0 {
```
Why is the check `len(r.ObjectMeta.OwnerReferences) == 0` not present for statefulsets/deployments?

Prateeknandle replied:
this was added in order to handle [comment](https://github.com/accuknox/discovery-engine/pull/681#pullrequestreview-1352921553), this case will not occur for sts and deployments.

Ankurk99 replied:
It can be possible that deployment are managed by `job` owner. I have added the checks for those cases.

Ankurk99 commented:
Hey @Prateeknandle, Can you please resolve the conflicts in the PR?

Prateeknandle committed:
resolving conflicts

Signed-off-by: Prateeknandle <prateeknandle@gmail.com>

Ankurk99 committed:
address review comments

- add check for owner reference in case of deployment and statefulsets
- show downloaded policy-template version

Signed-off-by: Ankur Kothiwal <ankur.kothiwal99@gmail.com>

vishnusomank reviewed (commented):
I think we should handle daemonsets as well in this. @Ankurk99 what do you think?

Ankurk99 committed:
add support for daemonsets

Signed-off-by: Ankur Kothiwal <ankur.kothiwal99@gmail.com>

nyrahul commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -199,3 +219,55 @@ func uniqueNsDeploy(deployName, deployNamespace string) *types.Deployment {
+		version, err := DownloadAndUnzipRelease()
+		if err != nil {
+			log.Error().Msgf("Unable to download %v", err.Error())
+			return nil
+		} else {
+			log.Info().Msgf("Downloaded version: %v", version)
+		}
```
```suggestion
			return nil
		}
		log.Info().Msgf("Downloaded version: %v", version)
```

seswarrajan reviewed (approved):
LGTM

nyrahul reviewed (approved):
Mostly good. Just one small fix.

nyrahul commented:
@vishnusomank , can you provide your approval?

Vyom-Yadav commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -144,7 +162,7 @@ func RecommendPolicyMain() {
 		}
 
-		for _, ns := range nsNotFilterSysPolicy {
+		for _, ns := range nsNotFilter {
 			if d.Namespace != ns {
 				generateHardenPolicy(d.Name, d.Namespace, d.Spec.Template.Labels)
```
We are already generating policies for `Deployment`s here. Why are we generating them again at line 150? It should be in a single place only.

Vyom-Yadav commented on src/recommendpolicy/recommendPolicy.go:
```diff
@@ -199,3 +219,54 @@ func uniqueNsDeploy(deployName, deployNamespace string) *types.Deployment {
+func GetHardenPolicy(deployments *v1.DeploymentList, replicaSets *v1.ReplicaSetList, statefulSets *v1.StatefulSetList, daemonSets *v1.DaemonSetList, nsNotFilter []string) []types.KnoxSystemPolicy {
+
+	var policies []types.KnoxSystemPolicy
+	if !isLatest() {
+		version, err := DownloadAndUnzipRelease()
+		if err != nil {
+			log.Error().Msgf("Unable to download %v", err.Error())
+			return nil
+		}
+		log.Info().Msgf("Downloaded version: %v", version)
+	}
```
This should not be clubbed with `GetHardenPolicy(...)`. Downloading and unzipping release is a separate process. Policy templates are required for both KubeArmor and Kyverno policies.
